src/etl/Source.hpp

```diff
@@ -48,40 +48,53 @@ namespace etl {
 template <
     typename GrpcSourceType = impl::GrpcSource,
     typename SubscriptionSourceTypePtr = std::unique_ptr<impl::SubscriptionSource>,
-    typename ForwardingSourceType = impl::ForwardingSource>
+    typename ForwardingSourceTypePtr = std::unique_ptr<impl::ForwardingSource>>
 class SourceImpl {
     std::string ip_;
     std::string wsPort_;
     std::string grpcPort_;
 
     GrpcSourceType grpcSource_;
     SubscriptionSourceTypePtr subscriptionSource_;
-    ForwardingSourceType forwardingSource_;
+    ForwardingSourceTypePtr forwardingSource_;
 
 public:
     using OnConnectHook = impl::SubscriptionSource::OnConnectHook;
     using OnDisconnectHook = impl::SubscriptionSource::OnDisconnectHook;
 
-    template <typename SomeGrpcSourceType, typename SomeForwardingSourceType>
-        requires std::is_same_v<GrpcSourceType, SomeGrpcSourceType> &&
-                     std::is_same_v<ForwardingSourceType, SomeForwardingSourceType>
+    template <typename SomeGrpcSourceType>
+        requires std::is_same_v<GrpcSourceType, SomeGrpcSourceType>
     SourceImpl(
         std::string ip,
         std::string wsPort,
         std::string grpcPort,
         SomeGrpcSourceType&& grpcSource,
         SubscriptionSourceTypePtr subscriptionSource,
-        SomeForwardingSourceType&& forwardingSource
+        ForwardingSourceTypePtr forwardingSource
     )
         : ip_(std::move(ip))
         , wsPort_(std::move(wsPort))
         , grpcPort_(std::move(grpcPort))
         , grpcSource_(std::forward<SomeGrpcSourceType>(grpcSource))
         , subscriptionSource_(std::move(subscriptionSource))
-        , forwardingSource_(std::forward<SomeForwardingSourceType>(forwardingSource))
+        , forwardingSource_(std::move(forwardingSource))
     {
     }
 
+    SourceImpl(SourceImpl const&) = delete;
+    SourceImpl(SourceImpl&&) = default;
+    SourceImpl&
+    operator=(SourceImpl const&) = delete;
+    SourceImpl&
+    operator=(SourceImpl&&) = default;
+
+    ~SourceImpl()
+    {
+        // subscriptionSource_ must die before forwardingSource_ because forwardingSource_ is used in callback
```

Comment:
Is there some way to avoid it? like for example use `shared_ptr` and have shared ownership of the `forwardingSource` maybe?